python/sglang/srt/layers/quantization/fp8.py

```diff
@@ -319,7 +316,25 @@ class Fp8MoEMethod(FusedMoEMethodBase):
         quant_config: The quantization config.
     """
 
-    def __init__(self, quant_config: Fp8Config):
+    def __new__(cls, *args, **kwargs):
```

Comment:
why need `__new__`?

Reply:
FusedMoEMethodBase needs to be inherited, but directly writing it as an import will cause circular dependencies. Currently, a dynamic approach is used to avoid this issue.

Reply:
Most other changes are what I spotted too, just `__new__` doesn't seem to be necessary?

Reply:
`__new__` is used here because we need to modify the class inheritance before instance creation. It's the only method that runs before `__init__` and allows us to control how the instance is created, letting us break the circular import by setting up inheritance at runtime rather than import time.

Reply:
If we use `apply` in `fp8.py`, and remove `apply` setting in `__init__.py`, should be simply ok?

Reply:
I didn't see `__new__` used before in sglang, vllm, so just curious, could be wrong though.

Reply:
Very good suggestion. Patching `get_quant_method` in `__init__` is because we want to centralize all patches related to quant in  `__init__.py`, including but not limited to Fp8, AWQ, GPTQ etc. What do you think? Thanks!

Reply:
ref https://www.geeksforgeeks.org/python-metaclass-__new__-method/

Reply:
May be we don't need to centralize them in `__init__.py`, if that can avoid circular?

Reply:
May you help complete this in the follow-up? I need to attend to other things after fixing this, as my bandwidth is quite limited recently.

Reply:
sure, may I have my other higher priority PR out and merge first?

Reply:
Thanks! If you are willing to handle some merge or rebase work for this PR after that PR is merged, there is no problem. However, I still recommend merging this one first as it is currently affecting users of the main branch.

Reply:
Okay, let's see, I didn't observe usability issue on main branch from https://github.com/sgl-project/sglang/pull/2388
It is also much needed on our side 😄

Reply:
ref https://github.com/sgl-project/sglang/issues/2386

Reply:
Thanks, let me take a look, my side of ROCm tests has got no complain, so worthy a check.

Reply:
```bash
python3 -c "from sglang.srt.layers.fused_moe_triton.fused_moe import fused_moe"
```

Reply:
I see it too, only used in benchmark scripts, so we will fix it, let me continue it tomorrow.

Reply:
Thanks! ref https://github.com/sgl-project/sglang/pull/2387#issuecomment-2525060144